llvm/lib/Target/AIE/AIEHazardRecognizer.cpp

```diff
@@ -49,51 +49,64 @@ void FuncUnitWrapper::setFormatInterface(const AIEBaseMCFormats *Formats) {
 
 bool FuncUnitWrapper::operator==(const FuncUnitWrapper &Other) const {
   return Required == Other.Required && Reserved == Other.Reserved &&
-         Slots == Other.Slots;
+         Slots == Other.Slots && MemoryBanks == Other.MemoryBanks;
 }
 
 void FuncUnitWrapper::dump() const {
   const char *const Digits = "0123456789";
   const char *const Spacer = "-|";
   const int Upper = std::numeric_limits<InstrStage::FuncUnits>::digits - 1;
-  dbgs() << "Req     : ";
-  for (int J = Upper; J >= 0; J--)
-    dbgs() << ((Required & (1ULL << J)) ? Digits[J % 10] : Spacer[J % 10 == 0]);
-  dbgs() << " Slots : ";
-  for (int J = 9; J >= 0; J--)
-    dbgs() << ((Slots & (1ULL << J)) ? Digits[J] : '-');
+
+  auto printFU = [&](const std::string &FUName, InstrStage::FuncUnits FU) {
+    dbgs() << FUName;
+    for (int J = Upper; J >= 0; J--)
+      dbgs() << ((Required & (1ULL << J)) ? Digits[J % 10]
+                                          : Spacer[J % 10 == 0]);
+  };
+  auto printResource = [&](const std::string &ResourceName, uint64_t Resource) {
+    dbgs() << ResourceName;
+    for (int J = 9; J >= 0; J--)
+      dbgs() << ((Resource & (1ULL << J)) ? Digits[J] : '-');
+  };
+
+  printFU("Req     : ", Required);
+  printResource(" Slots : ", Slots);
+  printResource(" Memorybanks : ", MemoryBanks);
   if (!Reserved)
     return;
-  dbgs() << "\n\t   Rsrv : ";
-  for (int J = Upper; J >= 0; J--)
-    dbgs() << ((Reserved & (1ULL << J)) ? Digits[J % 10] : Spacer[J % 10 == 0]);
+  printFU("\n\t   Rsrv : ", Reserved);
 }
 
 void FuncUnitWrapper::clearResources() {
   IssueCount = 0;
   Required = 0;
   Reserved = 0;
   Slots = 0;
+  MemoryBanks = 0;
 }
 bool FuncUnitWrapper::isEmpty() const {
-  return Required == 0 && Reserved == 0 && Slots == 0;
+  return Required == 0 && Reserved == 0 && Slots == 0 && MemoryBanks == 0;
 }
 
 void FuncUnitWrapper::blockResources() {
   Required = ~0;
   Reserved = ~0;
   Slots = ~0;
+  // Since the HW stalls in the event of memory bank conflicts, we don't need to
+  // block the resource. It is overly conservative if we block all memory banks.
```

Comment:
The comment is superfluous. It makes more sense to have MemoryBanks = ~0. The function is used to inform the scheduler that not a single resource should land there, and there's no use in wrecking your brain to find out whether MemoryBanks should be exempt. If there's a regression connected to this I would be interested, since I think it's hard to place a memory bank resource there without hitting any of the other blocks.

Reply:
We had a discussion regarding the same and had decided to be optimistically handle memory bank bits while we do scheduling across basic block. Today we have case where we force all the resources to be blocked while we do bottom up scheduling which is extremely pessimistic when it comes to bank resources.

In the following case when we schedule VST_CONV (which has an AS of 6) which translate into Memorybanks bit 1 to be used in the 7th cycle and thus we need to reserve Memory Bank in 7th cycle relative to emitting of the same, the 7th cycle now lies in the bot-scoreboard which is fully reserved thus leading to -6 delta cycle compared to -2 without AS. This behavior is caused due to blocking. 

![image](https://github.com/user-attachments/assets/1ee9e535-e005-4d6f-a9b1-ad95a30ba7ac)

Your (@martien-de-jong) opinions when we discussed this was 
` I agree, memory bank contention could be handled optimistically. I'm not sure how easy it is to separate recording the backs from recording the functional units. That should probably be accommodated in the top-level HazardRecognizer interfaces, perhaps using a state member of it.`

We should be completely optimistic here since the worst we can do is create a bank conflict and the HW support of stall mitigates it which is not possible for FU/Slots.

Reply:
Based on todays(31st July 2024) discussion we have concluded on not to reserve MemoryBank resources when doing inter basic block scheduling.